homeassistant/components/scheduler/schedule.py

```diff
@@ -0,0 +1,318 @@
+from datetime import datetime, timedelta
+import logging
+from typing import Awaitable, Callable, Dict, List, Optional, Tuple
+from uuid import uuid4
+
+from dateutil.rrule import rrule, rrulestr
+
+from homeassistant.const import CONF_ENTITY_ID, MATCH_ALL
+from homeassistant.core import Context, HomeAssistant, State, callback
+from homeassistant.helpers.event import (
+    async_track_point_in_time,
+    async_track_state_change,
+)
+from homeassistant.helpers.storage import Store
+import homeassistant.util.dt as dt_util
+
+_LOGGER = logging.getLogger(__name__)
+DOMAIN = "scheduler"
+
+STORAGE_VERSION = 1
+
+CONF_END_DATETIME = "end_datetime"
+CONF_RECURRENCE = "recurrence"
+CONF_SCHEDULE_ID = "schedule_id"
+CONF_START_DATETIME = "start_datetime"
+
+
+class ScheduleInstance:
+    """A class to represent an active instance of a schedule."""
+
+    def __init__(self, hass: HomeAssistant, entity_id: str):
+        """Initialize."""
+        self._affected_states: List[State] = []
+        self._async_state_listener: Optional[Callable[..., Awaitable]] = None
+        self._context = Context()
+        self._hass = hass
+        self.entity_id = entity_id
+
+    @callback
+    def async_cancel(self) -> None:
+        """Cancel the instance."""
+        if self._async_state_listener:
+            self._async_state_listener()
+            self._async_state_listener = None
+
+    async def async_revert(self) -> None:
+        """Restore the entities touched by the schedule."""
+        if not self._affected_states:
+            return
+
+        entities = {}
+        for state in self._affected_states:
+            data = {**state.attributes}
+            data["state"] = state.state
+            entities[state.entity_id] = data
+
+        await self._hass.services.async_call(
+            "scene",
+            "apply",
+            service_data={"entities": entities},
+            blocking=True,
+            context=self._context,
+        )
+
+        _LOGGER.info("Scheduler reverted scene: %s", self.entity_id)
+
+    async def async_trigger(self) -> None:
+        """Trigger the schedule's scene."""
+
+        @callback
+        def store_entity_if_in_context(
+            entity_id: str, old_state: State, new_state: State
+        ) -> None:
+            """Save prior states of an entity if it was triggered by this schedule."""
+            if new_state.context == self._context:
+                self._affected_states.append(old_state)
+
+        self._async_state_listener = async_track_state_change(
+            self._hass, MATCH_ALL, store_entity_if_in_context
+        )
+
+        await self._hass.services.async_call(
+            "scene",
+            "turn_on",
+            service_data={"entity_id": self.entity_id},
+            blocking=True,
+            context=self._context,
+        )
+
+        _LOGGER.info("Scheduler triggered scene: %s", self.entity_id)
+
+
+class Schedule:
```

Comment:
I think that we should store an `is_on` boolean in this class, controlled via an `async_turn_on` / `async_turn_off`. As long as the Schedule is on, it will keep track of time and instantiate a `ScheduleInstance` when necessary.